### PR TITLE
Added Filters Property to FileSystemWatcher class

### PR DESCRIPTION
**Pull request by Anipik**
Fixes https://github.com/dotnet/corefx/issues/29273

This property will help to use multiple filters for FileSystemWatcher.
The tests are built against netstandard, so have to add the tests to new files with .netcoreapp.

@JeremyKuhne  should I need to copy all the tests for all the operations like DirectoryChanged, DIrectoryDelete etc,. Currently I have just added a test for CreateDirectory Operation.

### CODE REVIEW

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 1]
```diff
@@ -697,4 +711,57 @@ private bool IsSuspended()
...
         }
     }
+
+    internal sealed class NormalizedFilterCollection : Collection<string>
+    {
+        public NormalizedFilterCollection()
+        {
+        }
+
+        public NormalizedFilterCollection(IList<string> list) : base(list)
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: As stated, you don't need this yet. We shouldn't have the constructor overload as discussed in the API review.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 2]
```diff
@@ -81,7 +79,7 @@ static FileSystemWatcher()
         public FileSystemWatcher()
         {
             _directory = string.Empty;
-            _filter = "*";
+            Filters.Add("*");
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: Don't do this. You shouldn't get `{ "*", "foo" }`  when adding a new filter. Add tests.

File: src/System.IO.FileSystem.Watcher/tests/System.IO.FileSystem.Watcher.Tests.csproj [hunk 3]
```diff
@@ -32,6 +32,7 @@
     <Compile Include="FileSystemWatcher.InternalBufferSize.cs" />
     <Compile Include="FileSystemWatcher.WaitForChanged.cs" />
     <Compile Include="FileSystemWatcher.unit.cs" />
+    <Compile Condition="$(Configuration.Split('-')[0]) == 'netcoreapp'" Include="FileSystemWatcher.unit.netcoreapp.cs" />
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.IO.FileSystem.Watcher/tests/System.IO.FileSystem.Watcher.Tests.csproj]: Presumably you'll need to add

```
      netcoreapp-Unix;
      netcoreapp-Windows_NT;
```

to the `Configurations.props`.

[the remark this answers is not in the document]
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/tests/System.IO.FileSystem.Watcher.Tests.csproj]: I already added that https://github.com/dotnet/corefx/pull/29315/files#diff-74516b96f1062e755007e73ab7adfaf7

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 4]
```diff
@@ -697,4 +711,57 @@ private bool IsSuspended()
...
+
+        protected override void SetItem(int index, string item)
+        {
+            if (string.IsNullOrEmpty(item))
+            {
+                // Skip the string compare for "*" since it has no case-insensitive representation that differs from
+                // the case-sensitive representation.
+                base.SetItem(index, "*");
+            }
+            else if (!string.Equals(this[index], item, PathInternal.StringComparison))
```
[the remark this answers is not in the document]
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: I kept this for https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.cs#L246

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 5]
```diff
@@ -138,6 +155,8 @@ public NotifyFilters NotifyFilter
             }
         }
 
+        public Collection<string> Filters { get; } = new NormalizedFilterCollection();
```
[the remark this answers is not in the document]
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: not having setter implies it as a readonly property, we cant use readonly attribute for properties.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs (hunk 2 above)
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: wont be we facing the same thing for ```public FileSystemWatcher(string path) : this(path, "*")``` should we change this also ?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 6]
```diff
@@ -198,12 +167,12 @@ public string Filter
         {
             get
             {
```
Comment (danmosemsft) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: I suggested that the getter throw off the is more than one in the list. Returning the first of many adds no value. Did API review mtg discuss this?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 7]
```diff
@@ -710,58 +695,12 @@ private bool IsSuspended()
...
-        }
 
-        protected override void ClearItems()
+        private class NormalizedFilterCollection : Collection<string>
         {
-            string filter = this[0];
-            base.ClearItems();
-            base.Add(filter);
-        }
+            protected override void InsertItem(int index, string item) => base.InsertItem(index, string.IsNullOrEmpty(item) || item == "*.*" ? "*" : item);
```
Comment (danmosemsft) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: Why are you bothering to add something to the list if I pass null or empty..you can leave list empty or throw

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs (hunk 6 above)
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: yeah i brought that up during the review mtg , Immo read ur comment but i forget the reason why we declined that

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs (hunk 7 above)
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: i will leave it empty in  null or empty case

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 8]
```diff
@@ -8,6 +8,7 @@
 using System.IO.Enumeration;
 using System.Threading;
 using System.Threading.Tasks;
+using System.Collections.ObjectModel;
```
Comment (MarcoRossignoli) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: nit: i read on [coding style](https://github.com/dotnet/corefx/blob/aca57537adf4015a8a69e0ae386970d45bd60c9a/Documentation/coding-guidelines/coding-style.md) should be sorted alphabetically right?

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+                
+                for (int i = 0; i < watcher.Filters.Count; i++)
+                {
+                    Assert.Equal(expected[i], watcher.Filters[i]);
+                }            
+            }
+        }
+
+        [Fact]
+        public void FileSystemWatchingMulptipleFilters()
```
Comment (danmosemsft) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: typo

**Comment by Anipik** [comment]
@Jeremykuhne can you please take another look at this ?

**Comment by Anipik** [comment]
@dotnet-bot test this please

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs (hunk 6 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: We didn't throw because we never threw before. It was a matter of consistency- you can't get this property to throw.

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 10]
```diff
@@ -98,21 +96,9 @@ public FileSystemWatcher(string path) : this(path, "*")
...
-            if (!Directory.Exists(path))
-                throw new ArgumentException(SR.Format(SR.InvalidDirName_NotExists, path), nameof(path));
-
+            CheckPathValidity(path);
             _directory = path;
-            _filter = filter ?? throw new ArgumentNullException(nameof(filter));
-
-            if (_filter == "*.*")
-                _filter = "*";
+            Filters.Add(filter); // ?? throw new ArgumentNullException(nameof(filter));
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: You need to throw for filter as it used to. Make sure we have tests that check the order of exceptions as well as all of the exceptions (e.g., if both are null you get "path", etc.)

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+                Assert.Empty(watcher.Filters);
+                Assert.NotNull(watcher.Filters);
+            }
+        }
+
+        [Fact]
+        public void InvalidOperationsOnFilters()
+        {
+            var watcher = new FileSystemWatcher();
+            Setup(watcher);
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: This helper makes it hard to follow what the test is doing. It isn't doing anything but add two items, just do it here. Same goes for other usages. If you want to assert what that helper does currently, add a specific test.

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+
+                action = () => File.Delete(secondFileName);
+                cleanup = () => File.Create(secondFileName).Dispose();
+                cleanup();
+
+                ExpectEvent(watcher, WatcherChangeTypes.Deleted, action, cleanup, expectedPaths);
+            }
+        }
+
+        private void EqualItems(Collection<string> firstCollection, Collection<string> secondCollection)
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: Why this instead of Assert.Equal?

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.cs [hunk 13]
```diff
@@ -237,13 +234,10 @@ public void FileSystemWatcher_Filter()
...
 
-            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows) || // expect no change for OrdinalIgnoreCase-equal strings
-                RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
+            if (!(PlatformDetection.IsFullFramework || PlatformDetection.IsOSX))
             {
-                // expect no change for OrdinalIgnoreCase-equal strings
-                // it's unclear why desktop does this but preserve it for compat
                 watcher.Filter = "ABC.DLL";
-                Assert.Equal("abc.dll", watcher.Filter);
+                Assert.Equal("ABC.DLL", watcher.Filter);
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.cs]: While this existing behavior is somewhat surprising, there isn't a strong reason to introduce a potential break here. I'd lean towards making it match. @pjanotti, what do you think?

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs (hunk 12 above)
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: because watcher.Filters is Derived Class  from Collection<T> which leads to failure of the test

Comment (JeremyKuhne) [reply src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: Which overload was it picking up?

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 14]
```diff
@@ -696,5 +695,30 @@ private bool IsSuspended()
...
             return _initializing || DesignMode;
         }
+
+        private class NormalizedFilterCollection : Collection<string>
+        {
+            protected override void InsertItem(int index, string item)
+            {
+                if (string.IsNullOrEmpty(item))
+                {
+                    ClearItems();
```
[the remark this answers is not in the document]
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: do we still want to clear the ```Filters``` when we set the ```Filter``` property ? what should be the output when we are setting the filter to  ```ABC.dll```  (The existing value is ```abc.dll```) ? should we clear the  ```Filters``` in that case as well ?

[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: > do we still want to clear the Filters when we set the Filter property ?

Yes

> what should be the output when we are setting the filter to ABC.dll (The existing value is abc.dll) 

As I mentioned, we should probably stay compatible here, i.e. case insensitive on Mac/Windows. Getting @pjanotti opinion as well.

[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: I discussed this more with @pjanotti offline. We think the right thing to do here is keep the value as given by the user. The only check/transform we should do is for null/empty. If you put "abc", then "ABC", you should get back "ABC".

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs (hunk 12 above)
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]:
```   
 System.IO.Tests.FileSystemWatcherFilterListTests.ClearFilters [FAIL]
        Assert.Equal() Failure
        Expected: Collection<String> []
        Actual:   NormalizedFilterCollection []
```
Similar is the case with ToString(). should we override ToString() ? is there any way to make the getType return collection<string> ? i tried using new keyword but its still returning ```NormalizedFilterCollection```

**Comment by Anipik** [comment]
@dotnet-bot test this please

File: src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs [hunk 15]
```diff
@@ -696,5 +695,23 @@ private bool IsSuspended()
...
+            {
+                base.InsertItem(index, string.IsNullOrEmpty(item) || item == "*.*" ? "*" : item);
+            }
+
+            protected override void SetItem(int index, string item)
+            {
+                base.SetItem(index, string.IsNullOrEmpty(item) || item == "*.*" ? "*" : item);
+            }
+
+            public override string ToString()
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs]: Don't override `ToString()`. There is no end-user value in a custom `ToString()`.

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs (hunk 12 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: Cast to `IEnumerable<string>` and it should work. This is the assert method you want: `public static void Equal<T>(IEnumerable<T> expected, IEnumerable<T> actual);` Don't write your own.

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+    public class FileSystemWatcherFilterListTests : FileSystemWatcherTest
+    {
+        [Fact]
+        public void DefaultFiltersValue()
+        {
+            var watcher = new FileSystemWatcher();
+            Assert.Equal(0, watcher.Filters.Count);
+            Assert.Empty(watcher.Filters);
+            Assert.NotNull(watcher.Filters);
+            EqualItems(new Collection<string> { }, watcher.Filters);
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: These should all be `Assert.Equal` with the expected being `new string [] {}`. That should get the right overload, if it doesn't, cast to `IEnumerable<string>()`- but only if you have to.

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+                string[] expectedPaths = new string[] { fileName, secondFileName };
+
+                watcher.Filters.Add(Path.GetFileName(fileName));
+                watcher.Filters.Add(Path.GetFileName(secondFileName));
+
+                Action action = () => File.Delete(fileName);
+                Action cleanup = () => File.Create(fileName).Dispose();
+                cleanup();
+
+                ExpectEvent(watcher, WatcherChangeTypes.Deleted, action, cleanup, expectedPaths);
```
[the remark this answers is not in the document]
Comment (Anipik) [reply src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: > Why pass both paths when you're only expecting one?

For some reason, in the second call to the ```ExpectEvent``` , expectedPath[0] is picking the first value only i.e. path to ```file``` rather than path to ```secondFile```.
I tried debugging it. The only statement in the second call to ```ExpectEvent``` where the ```expectedPath[0]``` is not ```secondFile```  suddenly is https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs#L84

File: src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.netcoreapp.cs [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        {
+            int attemptsCompleted = 0;
+            bool result = false;
+            FileSystemWatcher newWatcher = watcher;
+            while (!result && attemptsCompleted++ < attempts)
+            {
+                if (attemptsCompleted > 1)
+                {
+                    // Re-create the watcher to get a clean iteration.
+                    newWatcher = new FileSystemWatcher()
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.netcoreapp.cs]: Don't duplicate the entire method. Add a new helper that does only what is different (i.e. `FileSystemWatcher RecreateWatcher(FileSystemWatcher watcher)` or something like that).

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 19]
```diff
@@ -0,0 +1,485 @@
...
+                var watcher = new FileSystemWatcher(testDirectory.Path, "*.pdb");
+                watcher.Filters.Add("foo");
+                Assert.Equal(new string[] { "*.pdb", "foo" }, watcher.Filters);
+            }
+        }
+
+        [Fact]
+        public void FileSystemWatcher_File_Delete()
+        {
+            using (var testDirectory = new TempDirectory(GetTestFilePath()))
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: You already derive from FileCleanupTestBase, no need to use TempDirectory directly.  I've rewritten this test for clarity, hopefully this will help:

``` C#
[Fact]
public void FileSystemWatcher_File_Delete_MultipleFilters()
{
    // Check delete events against multiple filters

    DirectoryInfo directory = Directory.CreateDirectory(GetTestFilePath());
    FileInfo fileOne = new FileInfo(Path.Combine(directory.FullName, GetTestFileName()));
    FileInfo fileTwo = new FileInfo(Path.Combine(directory.FullName, GetTestFileName()));
    fileOne.Create().Dispose();
    fileTwo.Create().Dispose();

    using (var watcher = new FileSystemWatcher(directory.FullName))
    {
        watcher.Filters.Add(fileOne.Name);
        watcher.Filters.Add(fileTwo.Name);

        ExpectEvent(watcher, WatcherChangeTypes.Deleted, () => fileOne.Delete(), cleanup: null, expectedPath: fileOne.FullName);
        ExpectEvent(watcher, WatcherChangeTypes.Deleted, () => fileTwo.Delete(), cleanup: null, expectedPath: fileTwo.FullName);
    }
}
```

In the rewrite, I've grouped things together into phases. Setup (1), then execution (2).  In the original code setup happens interleaved with the execution. It becomes a mental exercise to understand if the setup has to be that way- is there an order dependency? In the update above it is clear there isn't.

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 20]
```diff
@@ -0,0 +1,460 @@
...
+            FileInfo fileTwo = new FileInfo(Path.Combine(directory.FullName, GetTestFileName()));
+            fileOne.Create().Dispose();
+            fileTwo.Create().Dispose();
+
+            using (var watcher = new FileSystemWatcher(directory.FullName))
+            {
+                watcher.Filters.Add(fileOne.Name);
+                watcher.Filters.Add(fileTwo.Name);
+
+                // using expectedPaths instead of expectedPath because otherwise in this call to ExpectEvent first lambda event handler (added during the last call) will fail.
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: You should fix the helper, not degrade the test to deal with a bug.
https://stackoverflow.com/questions/859936/unhooking-events-with-lambdas-in-c-sharp

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 21]
```diff
@@ -0,0 +1,460 @@
...
+            string directoryThree = Path.Combine(directory.FullName, GetTestFileName());
+
+            using (var watcher = new FileSystemWatcher(directory.FullName))
+            {
+                watcher.Filters.Add(Path.GetFileName(directoryOne));
+                watcher.Filters.Add(Path.GetFileName(directoryTwo));
+
+                ExpectEvent(watcher, WatcherChangeTypes.Created, () => Directory.CreateDirectory(directoryOne), cleanup: null, expectedPaths: new string[] { directoryOne, directoryTwo, directoryThree });
+                ExpectEvent(watcher, WatcherChangeTypes.Created, () => Directory.CreateDirectory(directoryTwo), cleanup: null, expectedPaths: new string[] { directoryOne, directoryTwo, directoryThree });
+                Assert.Throws<Xunit.Sdk.TrueException>(() => ExpectEvent(watcher, WatcherChangeTypes.Created, () => Directory.CreateDirectory(directoryThree), expectedPaths: new string[] { directoryOne, directoryTwo, directoryThree }));
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: This is really awkward. I'd create a helper here, say `ExpectNoEvent()` that does what you're doing here. Especially important as you don't want to keep retrying or have a long timeout in the negative case. Make sure your helper works by creating a failing test temporarily.

Comment (JeremyKuhne) [reply src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: Additionally, I would be consistent and test the negative for all your tests.

**Comment by Anipik** [comment]
@dotnet-bot test Windows x86 Release Build

File: src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs [hunk 22]
```diff
@@ -26,6 +25,11 @@ public abstract class FileSystemWatcherTest : FileCleanupTestBase
         public const int DefaultAttemptsForExpectedEvent = 3;       // Number of times an expected event should be retried if failing.
         public const int DefaultAttemptsForUnExpectedEvent = 2;     // Number of times an unexpected event should be retried if failing.
 
+        public static FileSystemEventHandler changeHandler;
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs]: You can't do this and have tests run concurrently. One possible answer is to have the `Watch*` methods return a tuple `(AutoResetEvent event, FileSystemEventHandler handler)` and unhook in the caller.

Even if this assembly is set to run sequentially, it would be bad to leave an avoidable time-bomb like this.

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs [hunk 23]
```diff
@@ -0,0 +1,476 @@
...
+            DirectoryInfo directoryTwo = Directory.CreateDirectory(Path.Combine(directory.FullName, GetTestFileName()));
+            DirectoryInfo directoryThree = Directory.CreateDirectory(Path.Combine(directory.FullName, GetTestFileName()));
+
+            using (var watcher = new FileSystemWatcher(directory.FullName))
+            {
+                watcher.Filters.Add(Path.GetFileName(directoryOne.FullName));
+                watcher.Filters.Add(Path.GetFileName(directoryTwo.FullName));
+
+                ExpectEvent(watcher, WatcherChangeTypes.Deleted, () => directoryOne.Delete(), cleanup: null, expectedPath: directoryOne.FullName);
+                watcher.Deleted -= deleteHandler;
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.netcoreapp.cs]: It doesn't make sense to unhook here. Move it where you have the context (i.e. `ExecuteAndVerifyEvents`).

File: src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs [hunk 24]
```diff
@@ -200,16 +214,17 @@ public static bool ExecuteAndVerifyEvents(FileSystemWatcher watcher, WatcherChan
         {
             bool result = true, verifyChanged = true, verifyCreated = true, verifyDeleted = true, verifyRenamed = true;
             AutoResetEvent changed = null, created = null, deleted = null, renamed = null;
-            string[] expectedFullPaths = expectedPaths == null ? null : expectedPaths.Select(e => Path.GetFullPath(e)).ToArray();
+            FileSystemEventHandler createHandler = null, changeHandler = null, deleteHandler = null;
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs]: It would be better to keep the tuple and name it. Less likely to make a mistake or introduce one if you keep the related info together.

``` C#
bool verifyRenamed = (expectedEvents & WatcherChangeTypes.Renamed) > 0;
(AutoResetEvent EventOccured, RenamedEventHandler Handler) renamed = default;

if (verifyRenamed)
    renamed = WatchRenamed(watcher);

action();

if (verifyRenamed)
{
    bool actual = renamed.EventOccured.WaitOne();
    watcher.Renamed -= renamed.Handler;

    // Validate
}
```

File: src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs [hunk 25]
```diff
@@ -161,32 +163,44 @@ public static void ExpectEvent(FileSystemWatcher watcher, WatcherChangeTypes exp
...
+        /// <summary>
+        /// Does verification that the given watcher will not throw exactly/only the events in "expectedEvents" when
+        /// "action" is executed.
+        /// </summary>
+        /// <param name="watcher">The FileSystemWatcher to test</param>
+        /// <param name="expectedEvents">All of the events that are expected to be raised by this action</param>
+        /// <param name="action">The Action that will trigger events.</param>
+        /// <param name="cleanup">Optional. Undoes the action and cleans up the watcher so the test may be run again if necessary.</param>
+        /// <param name="expectedPath">Optional. Adds path verification to all expected events.</param>
+        public static void ExpectNoEvent(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, Action cleanup = null, string expectedPath = null, int timeout = WaitForExpectedEventTimeout)
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs]: nit- `expectedEvents` should be `unexpectedEvents`, the comment should be updated.

**Comment by Anipik** [comment]
@JeremyKuhne is there any other feedback for this ?

File: src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.cs [hunk 26]
```diff
@@ -469,7 +464,7 @@ public void FileSystemWatcher_Windows_OnRenameGivesExpectedFullPath()
             using (var file = new TempFile(Path.Combine(dir.Path, "file")))
             using (var fsw = new FileSystemWatcher(dir.Path))
             {
-                AutoResetEvent eventOccurred = WatchRenamed(fsw);
+                AutoResetEvent eventOccurred = WatchRenamed(fsw).Item1;
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.unit.cs]: Put names on the return tuple to make this readable.

File: src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs [hunk 27]
```diff
@@ -30,11 +29,11 @@ public abstract class FileSystemWatcherTest : FileCleanupTestBase
         /// Watches the Changed WatcherChangeType and unblocks the returned AutoResetEvent when a
         /// Changed event is thrown by the watcher.
         /// </summary>
-        public static AutoResetEvent WatchChanged(FileSystemWatcher watcher, string[] expectedPaths = null)
+        public static (AutoResetEvent, FileSystemEventHandler) WatchChanged(FileSystemWatcher watcher, string[] expectedPaths = null)
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs]: As mentioned above, if you put names on these when you access them you can use the names. Unfortunately `ExecuteAndVerifyEvents` is structured in such a way that you have to initialize without the context so you can't leverage this. Other places you can so you should add them so you aren't doing `.Item1`.
